README.md
# Process Map

[![Build Status]][Travis] [![AppVeyor Build Status]][AppVeyor Link] [![Coverage status]][Coverage status link] [![Github Version]][Github Release Page] [![CRAN Version]][CRAN Link] [![Download Stats]][CRAN Link] [![License]][License Page]

[Build Status]: https://travis-ci.org/twang2218/pmap.svg?branch=master
[Travis]: https://travis-ci.org/twang2218/pmap

[CRAN Version]: http://www.r-pkg.org/badges/version/pmap
[CRAN Link]: https://cran.r-project.org/web/packages/pmap/index.html

[Github Version]: https://img.shields.io/github/release/twang2218/pmap.svg
[Github Release Page]: https://github.com/twang2218/pmap/releases

[Coverage status]: https://coveralls.io/repos/github/twang2218/pmap/badge.svg?branch=master
[Coverage status link]: https://coveralls.io/github/twang2218/pmap?branch=master

[Download Stats]: https://cranlogs.r-pkg.org/badges/grand-total/pmap?color=brightgreen

[AppVeyor Build Status]: https://ci.appveyor.com/api/projects/status/github/twang2218/pmap?branch=master&svg=true
[AppVeyor Link]: https://ci.appveyor.com/project/twang2218/pmap

[License]: https://img.shields.io/cran/l/pmap.svg
[License Page]: https://github.com/twang2218/pmap/blob/master/LICENSE.md

The goal of `pmap` is to provide the functionality of generating a process map from an event log with the user's preference.

## Installation

An older version of `pmap` is available on [CRAN](https://cran.r-project.org/web/packages/pmap/index.html), if you prefer to install this version, you can install it by:

```R
install.packages("pmap")
```

> However, based on the [CRAN policy](https://cran.r-project.org/web/packages/policies.html#Submission), a developer shouldn't submit a package to CRAN more than once within a month, therefore the [GitHub repo](https://github.com/twang2218/pmap/releases) will be the primary release channel, and the package will be submitted to CRAN only when it is possible. That is, the package version in CRAN can be a bit outdated.

To install the latest version, you can install `pmap` from GitHub directly:

```R
devtools::install_github("twang2218/pmap")
```

And, the users have the options to choose the installed version by specifying the version number in the command, as I [git tagged](https://github.com/twang2218/pmap/tags) each release:

```R
devtools::install_github("twang2218/pmap", ref = "v0.5.0")
```

## Usage

This is a demonstration of how to use `pmap` to create a process map from an event log. `sepsis` dataset in the `eventdataR` package will be used in the demonstration.

### Data preparation

Like any data analysis task, the first but the most important thing is to prepare our data.

Before the actual preparation steps, we should have a common ground on the terminology to be used later. There are mainly four terms, `Case`, `Activity`, `Category` and `Event`. The relation between the terms can be described as the following graph.

<p align="center"><img src="man/figures/relation_schema.svg" alt="process map without prune" height="300px" /></p>

And `eventlog` is a collection of `Event`. So, each row in the `eventlog` represents an `Event` object, and each `Event` contains several attributes, including:

* when - `timestamp`;
* who - `case_id`;
* what - `activity` and `category`;

Therefore `pmap` requires three mandatory fields and one optional field in the given `eventlog` data frame:

* `timestamp`: Represent the timestamps of the events when they occurred. The data type should be `POSIXct`. For the case of data type of `timestamp` is `character`, the package will attempt to convert the column to `POSIXct`, but it's just handy in some cases, it's better to make sure the `timestamp` column is in correct data type.
* `case_id`: Represent `Case` ID in the process paths. It is used to calculate the activity frequency or process performance.
* `activity`: Activity name.
* `category`(_optional since v0.4.0_): It is used to differentiate the grouped activities by different colors for a better visualization purpose. For example, the marketing activities with different purposes can be visualized by different colors, with one purpose each. If `category` is missing, the `activity` name will be used as `category` for coloring by default.

> `category` was previously called `event_type`, and required before `v0.3.2`. It is no longer necessary after `v0.4.0`.

Now, let's do the data preparation.

``` r
library(eventdataR)
library(dplyr)
library(pmap)

# Prepare the event log data frame
> eventlog <- eventdataR::sepsis %>%
    rename(
      timestamp = Complete_Timestamp,
      case_id = Case_ID,
      activity = Activity
    ) %>%
    select(timestamp, case_id, activity) %>%
    na.omit()
```

Check `eventlog` data frame structure.

```R
> head(eventlog)
# A tibble: 6 x 3
  timestamp           case_id activity
  <dttm>              <chr>   <chr>
1 2014-10-22 11:15:41 A       ER Registration
2 2014-10-22 11:27:00 A       Leucocytes
3 2014-10-22 11:27:00 A       CRP
4 2014-10-22 11:27:00 A       LacticAcid
5 2014-10-22 11:33:37 A       ER Triage
6 2014-10-22 11:34:00 A       ER Sepsis Triage
> str(eventlog)
Classes ‘tbl_df’, ‘tbl’ and 'data.frame':       15190 obs. of  3 variables:
 $ timestamp: POSIXct, format: "2014-10-22 11:15:41" "2014-10-22 11:27:00" ...
 $ case_id  : chr  "A" "A" "A" "A" ...
 $ activity : chr  "ER Registration" "Leucocytes" "CRP" "LacticAcid" ...
 - attr(*, "na.action")=Class 'omit'  Named int [1:24] 442 443 444 445 446 447 448 449 450 451 ...
  .. ..- attr(*, "names")= chr [1:24] "442" "443" "444" "445" ...
```

### Create a process map

You can create a process map from the `eventlog` directly by running only one command:

```R
# Create process map
> p <- create_pmap(eventlog)
# Render the process map
> render_pmap(p)
```

The result will be shown in `Viewer` window if you're using R Studio, or in a new browser window if you're running the code from a Terminal.

<p align="center"><img src="man/figures/example.prune_edges.none.svg" alt="process map without prune" height="600px" /></p>

### Prune the process map

As you can see, the above result is a bit messy, however, we can prune some edges with smaller volume to simplify the process map. It is a better way to find the common paths in the process.

```R
p %>% prune_edges(0.5) %>% render_pmap()
```

<p align="center"><img src="man/figures/example.prune_edges.edges.svg" alt="process map without prune" height="500px" /></p>

It's better, but we can improve it even better by pruning some not very important nodes as well.

```R
> p %>% prune_nodes(0.5) %>% prune_edges(0.5) %>% render_pmap()
```

Or, if you want a more interactive approach, you can start a Shiny server app with a slide bar for pruning the nodes and/or edges by a certain percentage. Just be careful, the more the edges and nodes, the slower the process will be. Let's keep `50%` nodes and `50%` edges in our example:

```R
> render_pmap_shiny(p, nodes_prune_percentage = 0.5, edges_prune_percentage = 0.5)
```

<p align="center"><img src="man/figures/example.prune_edges.both.svg" alt="cleaner process map" height="500px" /></p>

### Expand the loop

The above process map is great to find the valuable insights as we can immediately observe something very interesting insight from the map: the loop between `CRP` and `Leucocyte`. Is this because of a small group of cases repeatedly went through these two steps many many times? or is this because most cases went through the loop just a few times? To answer the question, we can expand the loop by distinct the repeated activity.

```R
> p <- create_pmap(eventlog, distinct_repeated_activities = TRUE)
```

By this way, each new activity name will be attached with the occurrence sequence number of the activity in the path, so the same activity occurs multiple times in the path will have a different name, which means different nodes in the final map. The newly generated the process map will be much more complex than before, so we need prune it further.

```R
> p %>% prune_nodes(0.5) %>% prune_edges(0.8) %>% render_pmap()
```

<p align="center"><img src="man/figures/example.distinct_repeated_activities.svg" alt="process map with distinct repeated activities" height="700px" /></p>

It's interesting to see that there isn't much connection between first time `CRP (1)` and `Leucocyte (1)`, however, the back and forth happened after `Admission NC`.

### Time is the key

Are this back and forth activity loop because of some kind of regular check after `Admission NC`? We are not sure from previous process map, because we don't know how long between each activity occurred. By default, the edge label will be the number of cases went between the two connected activities. We can change it for the duration of those connected activities to understand more about the process in a timely manner.

As there are multiple cases went through the path, we need to decide how to summarize the duration, such as:

* the maximum duration
* the minimum duration
* the mean duration
* the median duration

We can specify the kind of duration by given an `edge_label` argument to the `create_pmap()`.

```R
p <- create_pmap(eventlog, edge_label = "mean_duration")
p %>% prune_nodes(0.5) %>% prune_edges(0.8) %>% render_pmap()
```

or, it can be changed after the process map created by using `adjust_edge_label()` function.

```R
p <- adjust_edge_label(p, label = "mean_duration")
render_pmap(p)
```

<p align="center"><img src="man/figures/example.distinct_repeated_activities.mean_duration.svg" alt="process map with distinct repeated activities with mean duration" height="700px" /></p>

By adding the duration between each path into the process map, we can eliminate the problem immediately, as it's not a loop. `Leucocyte` almost always occurred immediately after `CRP` occurred, but not the otherwise. It means `CRP` and `Leucocyte` occurred together in the same sequence, might belong to a blood test pack, and the patients will be tested regularly after `Admission NC`.

And we can also discover the patients would normally be released 2 days after `CRP` and `Leucocyte` test. It might because the test results came back ok, the patients will be released after 2 days observation without any further issue.

We can get that information from the process map clearly.

### Persistent the result

If you're happy with the result, you can save the process map to a PDF or other file format by replace `render_pmap()` with `render_pmap_file()`.

```R
p <- create_pmap(eventlog, edge_label = "mean_duration")
p %>% prune_nodes(0.5) %>% prune_edges(0.8) %>% render_pmap_file("sepsis_process_map.pdf")
```


NEWS.md
# pmap 0.6.0

Besides the count of the edge on the label, the duration stats are also calculated to be shown. Add `distinct_repeated_events` to be able to expand the repeated events to have better internal understanding. Add ability to use external `dot` program to render the graph, it will be faster in some cases.

* af20391 Fix 'UQ()' issue on dependencies.
* 0953e15 Fix figures naming issue.
* e1daa8c Add term schema relation graph
* 6523f17 Rename `activity_category` to just `category`
* 1929113 Change the term from `event` to `activity`.
* 384de85 Rename term `customer` to `case`
* 0c63555 Update the graph files and add more readme content to reflect latest update
* 702c526 Add `adjust_edge_label()` function
* a35ec17 Add huge process map warning for rendering
* 1fba54b Add `use_external_dot` argument to `render_pmap_file()`
* 089a958 Add more ignores to `.gitignore` and `.Rbuildignore`
* 1c17eb4 Add `dot` file support for `render_pmap_file()`
* 934d5d4 Add `distinct_repeated_events` argument to `create_pmap()`
* 095444c Remove node's color gradient as there is a compatibility issue in Viz.js
* de905fc Enhance color palette for event categories to 80 colours, and won't fail if ask more
* 64a089e Rename event_type to event_category to avoid confusion
* fa527d7 Add `*_prune_percentage` to `render_pmap_shiny()` for initial prune value settings.
* 84e656c Add ability to `render_pmap_file()` to guess file format
* 15ae4fa Handle the SPACE padding case
* 1530337 Convert `eventlog$timestamp` to `POSIXct` automatically if it's not `POSIXct` yet.
* c67f6c8 Add `median_duration` for edge label and more test cases.
* 8702685 Fix the edge label too close to the edge problem by padding some spaces
* dc79c7c Add duration support in edge label
* d26c5a5 Add Github version and license badges


# pmap 0.5.0

This update is mainly about performance improvements. With many of these tuning/refactoring, `0.5.0` is much faster than `0.4.0`. It can be thousands times faster for large dataset, such as event log contains several millions records.

Although the `create_pmap()` is pretty fast, the rendering part is still on V8's Javascript engine, which can be pretty slow for a graph with many nodes and edges. `prune_nodes()` and `prune_edges()` can be used to reduce the final graph's complexity.

* 94b1dd2 Improve slicing performance
* a4c72a0 Improve sorting performance by 2x or more.
* 073c2a4 Rewrite `generate_edges()` to improve performance
* e899f7d Add Installation section in `README.md` and change the example back for better compatibility.
* 8a4a096 Improve `get_colors()` performance by remove `ggsci` from dependencies
* 395cc42 Improve `create_pmap_graph()` and `generate_edges()` performance
* 3c98ac5 Remove data.table from required dependencies
* e7f4f14 Avoid running slow example
* a3788ec Refactor `get_colors()` function to make it more efficient.

# pmap 0.4.0

* 5e22bd8 Add `render_pmap_file()` function.
* 64d4a08 Use `match.arg()` for `prune_nodes()` argument `rank`
* c7537b6 Remove internal functions from the docs
* c2e74bc Reorder the prune sequence in `README.md` example
* b0d0880 Remove the dependency of eventdataR which is not necessary
* 0f97cf8 Update the `README.md` to use sepsis dataset as the example
* 63de70d Add `clean_graph()` to remove nodes without edge
* 63d6760 Explicitly using package name in the external function call.
* b26447d Remove the uncertainty by explicitly using `round()`.
* 6624434 Add ability to handle the `event_type` missing `eventlog` case.
* 12f1319 Add `render_pmap_shiny()` function
* 43fa193 Add `weight` attribute to the edge with projected `amount` value.
* ed365e3 Add test case with prune for `create_pmap()`
* dacdbdd Clean up `apply_node_color.R`, extract color-related code.
* 8423941 Add `adjust_node_style()` function
* a9cc7e6 Adjust edge style by simply project the `edges$amount` value to expected range.
* 76094cd Adjust the node size based on the amount within size range from 10 to 20
* d05d8fb Add randomness to the weight of each event names
* d91c3e0 Add `amount` in the node label.
* d70731a Remove the randomness of color selection, use predefined sequence.
* 1ea5a05 Add `apply_node_color()` function to dynamic assign the node's color based on the type
* 714dadd Add `CODE_OF_CONDUCT.md` and fix typos in `DESCRIPTION`.

# pmap 0.3.2

* Added a `NEWS.md` file to track changes.
* Add bug report url and more test environments in `cran-comments.md`

# pmap 0.3.1

* Replace PNG format with SVG format.
* Fix figures file namepackage.

# pmap 0.3.0

* Fix typos in `README.md` example
* Rename `create_pmap_from_eventlog()` to `create_pmap()`
* Rename `create_pmap()` to `create_pmap_graph()`
* Rename `generate_random_eventlog()` to `generate_eventlog()` and remove the `helper.R`
* Add more examples to the documents and fix typos
* Add release target to Makefile

# pmap 0.2.0

* Add `cran-comments.md`
* Fix typos
* Add example code to `README.md`
* Add more docs for `prune_edges()` and others
* Add more docs for `create_pmap_from_eventlog()`
* Add random eventlog generator function `generate_random_eventlog()`
* Add random eventlog generator function `generate_random_eventlog()`
* Add `create_pmap_from_eventlog()` example with graph
* Fix a bug that the target event shouldn't be the start point
* Add ability to handle the empty `target_types`, for that case, every paths count, rather than only the paths reaches the target

# pmap 0.1.0

* First release.

R/process_graph.R
# make 'R CMD check' happy
utils::globalVariables(c(
  # get_nodes_from_event_logs()
  "event_name",
  "is_target",
  "name",
  # get_edges_from_event_logs()
  "customer_id",
  "timestamp",
  "from",
  "to"
))

#' @title Produce nodes from event logs
#'
#' @param event_logs Event logs
#' @description `event_logs` should be a `data.frame` or `data.table`, which contains, at least, following columns:
#'
#'  * `event_name`: event name. (`character`)
#'  * `is_target`: whether it's the final stage. (`logical`)
#'
#' `get_nodes_from_event_logs()` will generate the node list from the given `event_logs` for the graph purpose.
#' However, it is very basic, so, in most cases, you might want to provide your own node list to the
#' `create_event_graph()` function.
#' @return a `data.frame` of nodes
#' @importFrom dplyr      %>%
#' @importFrom dplyr      select
#' @importFrom dplyr      distinct
#' @importFrom dplyr      rename
#' @importFrom dplyr      mutate
#' @importFrom dplyr      arrange
#' @export
get_nodes_from_event_logs <- function(event_logs) {
  nodes <- event_logs %>%
      select(event_name, is_target) %>%
      distinct() %>%
      rename(name = event_name) %>%
      mutate(name = as.character(name), percentage = 0.5) %>%
      arrange(name)
  return(nodes)
}

#' @title Produce edges from event logs
#' @param event_logs Event logs
#' @description `event_logs` should be a `data.frame` or `data.table`, which contains, at least, following columns:
#'
#'  * `timestamp`: timestamp column which indicates when event happened. (`POSIXct`)
#'  * `customer_id`: cutomer identifier. (`character`)
#'  * `event_name`: event name. (`character`)
#'  * `is_target`: whether it's the final stage. (`logical`)
#' @return a `data.frame` of edges
#' @importFrom dplyr        %>%
#' @importFrom dplyr        arrange
#' @importFrom dplyr        group_by
#' @importFrom dplyr        summarize
#' @importFrom dplyr        ungroup
#' @importFrom dplyr        mutate_if
#' @importFrom dplyr        n
#' @importFrom data.table   data.table
#' @importFrom data.table   rbindlist
#' @export
get_edges_from_event_logs <- function(event_logs) {
  # sort by customer_id and timestamp
  event_logs <- data.table(event_logs) %>% arrange(customer_id, timestamp)

  edges <- list()
  temp_edges <- list()
  previous <- NA

  # `apply()` is about up to 10x faster than `for-loop`
  apply(event_logs, 1, function(current) {
    if (!any(is.na(previous))) {
      # pre-process
      if (current["customer_id"] != previous["customer_id"]) {
        # it's a new customer, so set it to 'previous' and continue to the next record
        previous <<- current
        # clear temp_edges
        temp_edges <<- list()
        return
      }

      edge <- list(
        from = previous["event_name"],
        to = current["event_name"],
        customer_id = current["customer_id"]#,
        # duration = current["timestamp"] - previous["timestamp"]
      )
      temp_edges[[length(temp_edges) + 1]] <<- edge
    }

    # post-process
    if (current["is_target"] != FALSE) {
      # put temp_edges to final edges set
      if (length(temp_edges) > 0) {
        for (i in 1:length(temp_edges)) {
          edges[[length(edges) + 1]] <<- temp_edges[[i]]
        }
      }
      # target events should not be the starting point
      previous <<- NA
      # clear temp_edges
      temp_edges <<- list()
    } else {
      previous <<- current
    }
  })

  edges <- rbindlist(edges) %>% 
    group_by(from, to) %>%
    # Add attributes: `value` => count
    summarize(value = n()) %>%
    ungroup() %>%
    mutate_if(is.factor, as.character) %>%
    arrange(from, to)

  return(edges)
}

#' @title Create the event graph by given nodes and edges.
#' @param nodes Event list
#' @param edges Event transform list
#' @description
#' `nodes` should be a `data.frame` containing following columns:
#'   * `name`: Nodes name, will be used as label. (`character`)
#'   * `is_target`: Whether it's final stage. (`logical`)
#'   * `percentage`: The percentage of customer affected by the given event. (`numeric`)
#'
#' `edges` should be a `data.frame` containing following columns:
#'   * `from`: the begining event of the edge. (`character`)
#'   * `to`: the ending event of the edge (`character`)
#'   * `value`: How many of customer affected by the given event. (`numeric`)
#' @importFrom dplyr        %>%
#' @importFrom dplyr        mutate
#' @importFrom dplyr        mutate_if
#' @importFrom DiagrammeR   create_graph
#' @importFrom DiagrammeR   add_nodes_from_table
#' @importFrom DiagrammeR   add_edges_from_table
#' @importFrom DiagrammeR   add_global_graph_attrs
#' @importFrom DiagrammeR   set_node_attrs
#' @importFrom DiagrammeR   set_edge_attrs
#' @importFrom DiagrammeR   select_nodes_by_degree
#' @importFrom DiagrammeR   get_selection
#' @importFrom DiagrammeR   delete_nodes_ws
#' @export
create_event_graph <- function(nodes, edges) {
  # print("Converting factor to character [nodes]...")
  nodes <- nodes %>%
    mutate_if(is.factor, as.character) %>%
    mutate(
      index = 1:nrow(nodes),
      tooltip = get_attrs_desc(nodes)
    )
  # print(str(nodes))

  # print("Converting factor to character [edges]...")
  edges <- edges %>%
    mutate_if(is.factor, as.character) %>%
    mutate(
      tooltips = get_attrs_desc(edges)
    )
  # print(str(edges))
  # print("create_graph()")
  p <- create_graph()

  # print("add_nodes_from_table()")
  p <- p %>%
    add_nodes_from_table(
      table = nodes,
      label_col = "name"
    )

  # print("add_edges_from_table()")
  p <- p %>% add_edges_from_table(
      table = edges,
      from_col = "from",
      to_col = "to",
      ndf_mapping = "name")

  # print("add_global_graph_attrs()")
  p <- p %>% 
    # graph [ layout = "dot" ]
    add_global_graph_attrs(attr_type = "graph", attr = "layout", value = "dot") %>%
    # node [...]
    add_global_graph_attrs(attr_type = "node", attr = "fixedsize", value = "false") %>%
    add_global_graph_attrs(attr_type = "node", attr = "shape", value = "box") %>%
    add_global_graph_attrs(attr_type = "node", attr = "style", value = "filled,rounded") %>%
    add_global_graph_attrs(attr_type = "node", attr = "gradientangle", value = "90") %>%
    # edge [...]
    ## grey900(#212121)
    add_global_graph_attrs(attr_type = "edge", attr = "color", value = "#21212180") %>%
    ## grey900(#212121)
    add_global_graph_attrs(attr_type = "edge", attr = "fontcolor", value = "#212121")

  # print("set_node_attrs()")
  p <- p %>%
    # node attributes
    ## grey900(#212121)
    set_node_attrs(node_attr = "fontcolor", values = "#212121") %>%
    ## lightBlue900(#01579B)
    set_node_attrs(node_attr = "color", values = "#01579B") %>%
    ## lightBlue100(#B3E5FC) => lightBlue50(#E1F5FE)
    set_node_attrs(node_attr = "fillcolor", values = "#B3E5FC:#E1F5FE") %>%
    set_node_attrs(node_attr = "fontsize", values = (nodes$percentage * 15 + 5)) %>%
    set_node_attrs(node_attr = "label", values = nodes$name) %>%
    set_node_attrs(node_attr = "tooltip", values = nodes$tooltip)
    
  # print("set_edge_attrs() for target nodes")
  target_ids <- nodes[nodes$is_target, "index"]
  p <- p %>%
    ## deepOrange900(#BF360C)
    set_node_attrs(node_attr = "color", values = "#BF360C", nodes = target_ids) %>%
    ## deepOrange100(#FFCCBC):deepOrange50(#FBE9E7)
    set_node_attrs(node_attr = "fillcolor", values = "#FFCCBC:#FBE9E7", nodes = target_ids) %>%
    set_node_attrs(node_attr = "fontsize", values = (nodes$percentage * 10 + 15), nodes = target_ids)

  # print("set_edge_attrs()")
  p <- p %>%
    # Edge attributes
    set_edge_attrs(edge_attr = "penwidth", values = log10(edges$value) + 1) %>%
    set_edge_attrs(edge_attr = "label", values = edges$value) %>%
    set_edge_attrs(edge_attr = "tooltip", values = edges$tooltips) %>%
    set_edge_attrs(edge_attr = "labeltooltip", values = edges$tooltips)

  # remove node without edges
  zero_degree_nodes <- p %>% select_nodes_by_degree("deg == 0") %>% get_selection()
  if (!is.na(zero_degree_nodes) && length(zero_degree_nodes) > 0) {
    p <- p %>% select_nodes_by_degree("deg == 0") %>% delete_nodes_ws()
  }

  return(p)
}
